pkg/frontend/querymiddleware/codec.go

```diff
@@ -143,12 +145,25 @@ func newPrometheusCodecMetrics(registerer prometheus.Registerer) *prometheusCode
 }
 
 type prometheusCodec struct {
-	metrics *prometheusCodecMetrics
+	metrics               *prometheusCodecMetrics
+	requestProtobufFormat bool
 }
 
-func NewPrometheusCodec(registerer prometheus.Registerer) Codec {
+type format interface {
+	EncodeResponse(resp *PrometheusResponse) ([]byte, error)
+	DecodeResponse([]byte) (PrometheusResponse, error)
+	Name() string
+}
+
+var knownFormats = map[string]format{
+	jsonMimeType:                  jsonFormat{},
+	mimirpb.QueryResponseMimeType: protobufFormat{},
+}
+
+func NewPrometheusCodec(registerer prometheus.Registerer, cfg Config) Codec {
```

Comment:
[design] I suggest to take in input only the preferred response format instead of the whole Config.

Reply:
Fixed in ed988dc.